reasoner/resolution/resolver/BoundConclusionResolver.java

```diff
@@ -0,0 +1,377 @@
+        ConclusionRequestState<? extends Concludable<?>> requestState = this.requestStates.get(fromUpstream.first());
+        LOG.trace("{}: received materialisation response: {}", name(), response);
+        Optional<Map<Identifier.Variable, Concept>> materialisation = response.materialisation();
+        materialisation.ifPresent(m -> requestState.newMaterialisation(response.partialAnswer(), m));
+        sendAnswerOrSearchDownstreamOrFail(fromUpstream.first(), requestState, fromUpstream.second());
+        ConclusionRequestState.WaitedMaterialisations state = requestState.waitedMaterialisations();
+        state.decrement();
+        state.nextWaitingRequest().ifPresent(w -> {
+            sendAnswerOrSearchDownstreamOrFail(w.first(), requestState, w.second());
+        });
+    }
+
+    protected Pair<Request, Integer> fromUpstream(Materialiser.Request toDownstream) {
+        assert materialiserRequestRouter.containsKey(toDownstream);
+        return materialiserRequestRouter.remove(toDownstream);
+    }
+
+    private void sendAnswerOrSearchDownstreamOrFail(Request fromUpstream, ConclusionRequestState<?> requestState, int iteration) {
+        Optional<? extends AnswerState.Partial<?>> upstreamAnswer = requestState.nextAnswer();
+        if (upstreamAnswer.isPresent()) {
+            answerToUpstream(upstreamAnswer.get(), fromUpstream, iteration);
+        } else if (!requestState.isComplete() && requestState.downstreamManager().hasDownstream()) {
+            requestFromDownstream(requestState.downstreamManager().nextDownstream(), fromUpstream, iteration);
+        } else if (requestState.waitedMaterialisations().waiting()) {
+            requestState.waitedMaterialisations().addWaitingRequest(fromUpstream, iteration);
+        } else {
+            requestState.setComplete();
+            failToUpstream(fromUpstream, iteration);
+        }
+    }
+
+    @Override
+    protected void receiveFail(Response.Fail fromDownstream, int iteration) {
+        LOG.trace("{}: received Fail: {}", name(), fromDownstream);
+        if (isTerminated()) return;
+
+        Request toDownstream = fromDownstream.sourceRequest();
+        Request fromUpstream = fromUpstream(toDownstream);
+        ConclusionRequestState<?> requestState = this.requestStates.get(fromUpstream);
+
+        if (iteration < requestState.iteration()) {
+            // short circuit old iteration fail messages to upstream
+            failToUpstream(fromUpstream, iteration);
+            return;
+        }
+
+        requestState.downstreamManager().removeDownstream(fromDownstream.sourceRequest());
+        sendAnswerOrSearchDownstreamOrFail(fromUpstream, requestState, iteration);
+    }
+
+    @Override
+    protected void initialiseDownstreamResolvers() {
+        throw TypeDBException.of(ILLEGAL_STATE);
+    }
+
+    private ConclusionRequestState<?> createRequestState(Request fromUpstream, int iteration) {
```

Comment:
for conclusion rather than concludable the code paths don't justify a whole new class to manage each possibility. They're basically one method different.